includes/Story_Renderer.php

```diff
@@ -0,0 +1,178 @@
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function replace_html_start_tag( $content ) {
+		if ( preg_match( self::HTML_STRUCTURE_HTML_START_TAG, $content, $matches ) ) {
+			$replacement = sprintf( '<html amp %s>', get_language_attributes( 'html' ) );
+			$result      = preg_replace( self::HTML_STRUCTURE_HTML_START_TAG, $replacement, $content, 1 );
+
+			if ( null !== $result ) {
+				return $result;
+			}
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Returns the full HTML <head> markup for a given story besides boilerplate.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function get_html_head_markup() {
+		ob_start();
+		?>
+		<style amp-custom>
+			.page-background-area, .page-safe-area {
+				position: absolute;
+				overflow: hidden;
+				margin: auto;
+				top: 0;
+				bottom: 0;
+				left: 0;
+				right: 0;
+			}
+
+			.page-background-area img, .page-background-area video {
+				object-fit: cover;
+			}
+
+			.wrapper {
+				position: absolute;
+				overflow: hidden;
+			}
+
+			.fill {
+				position: absolute;
+				top: 0;
+				left: 0;
+				right: 0;
+				bottom: 0;
+				margin: 0;
+			}
+		</style>
+		<meta name="generator" content="<?php printf( 'Web Stories %s', esc_attr( WEBSTORIES_VERSION ) ); ?>" />
+		<?php
+
+		/**
+		 * Prints scripts or data in the head tag on the front end.
+		 *
+		 * @since 1.3
+		 */
+		do_action( 'web_stories_story_head' );
+
+		return (string) ob_get_clean();
+	}
+
+	/**
+	 * Replaces markers in HTML <head> with dynamic content.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function replace_html_head( $content ) {
+		$start_tag = '<meta name="web-stories-replace-head-start"/>';
+		$end_tag   = '<meta name="web-stories-replace-head-end"/>';
+
+		$start_tag_pos = strpos( $content, $start_tag );
+		$end_tag_pos   = strpos( $content, $end_tag );
+
+		if ( false !== $start_tag_pos && false !== $end_tag_pos ) {
+			$end_tag_pos += strlen( $end_tag );
+			$content      = substr_replace( $content, $this->get_html_head_markup(), $start_tag_pos, $end_tag_pos - $start_tag_pos );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Adds square, and landscape poster images to the <amp-story>.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function add_poster_images( $content ) {
+		$poster_images = Media::get_story_meta_images( $this->post );
+		unset( $poster_images['poster-portrait'] ); // Already exists.
+		foreach ( $poster_images as $attr => $url ) {
+			$attr_markup = sprintf( '%1$s-src="%2$s"', $attr, $url );
+			$content     = str_replace( 'poster-portrait-src=', $attr_markup . ' poster-portrait-src=', $content );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Renders the story.
+	 *
+	 * @return string The complete HTML markup for the story.
+	 */
+	public function render() {
+		$markup = $this->post->post_content;
+		$markup = $this->replace_html_start_tag( $markup );
```

Comment:
Is there a way to skip this, too? (I might be missing something being late to the review party but) it looks like everything between the meta tags is always replaced which seems to make the `CustomCSS` obsolete.

Also, doesn't this always remove the `<link rel="canonical" href={story.link} />`?

Reply:
> Also, doesn't this always remove the `<link rel="canonical" href={story.link} />`?

It would _override_ it, if that's what you mean.

As for the CSS, there is currently no need to modify it server-side, so I'll probably just change the marker there so it's not touched.

Reply:
Well, yes, it seems that it would _always_ override it. I guess my point is -- if there is no way how to stop overriding this link via `replace_html_head`, then what is the reason for adding it to the JS file at all?

Doesn't `replace_html_head` always replace everything between `web-stories-replace-head-start` and `web-stories-replace-head-end` currently?

And this might also mean removing the canonical link as well if there's no replacement.

Anyway, let me test this locally -- does everything work as expected for you and the canonical link is displayed as expected?

Reply:
> Well, yes, it seems that it would always override it. I guess my point is -- if there is no way how to stop overriding this link via `replace_html_head`, then what is the reason for adding it to the JS file at all?

To produce valid AMP output that can be used outside of PHP context.

> Doesn't `replace_html_head` always replace everything between `web-stories-replace-head-start` and `web-stories-replace-head-end` currently?

Correct.

> And this might also mean removing the canonical link as well if there's no replacement.

When would there be no replacement?

The idea here is that a canonical link is always present (valid AMP, can be used for previews outside WordPress, etc.), but also render it dynamically in PHP (e.g. when an SEO plugin wants to modify it or something).

So far it works as expected locally.

Reply:
I'm actually getting a missing canonical link invalid AMP error right now when testing.

Should `rel_canonical` added via `web_stories_story_head` be adding the canonical link in PHP?

Reply:
Okay, looks like the error happens in Draft mode only.